Added Instance Counts and Per-User Instance stats for fs.inotify.max_user_instances

bdowling opened the pull request:
Just a quick extension of your script to have instance stats, since this is useful to know when fs.inotify.max_user_instances limits are being hit.

Cheers!

The per-user stats are slightly off from the TOTAL INSTANCE COUNT, but I think that's due to the fact the script is counting its own sub processes, alternatively I suppose it could use the data it already collected to count each user's total.

bdowling committed:
Added Instance Counts and Per-User Instance stats for fs.inotify.max_user_instances

fatso83 commented:
Hi, thanks for this! I see that it results in a 9x increase in kernel system time (55 ms vs 489 ms) and 3x increase in wall time on my computer, so I don't think I can merge it as a default setting:

```
real	0m1,103s
user	0m0,636s
sys	0m0,489s
```

vs the old time
```
real	0m0,354s
user	0m0,239s
sys	0m0,055s
```
There are [people running it on big iron with literally hundreds of cores ](https://github.com/fatso83/dotfiles/commit/12831d1df66ea49dd585e6f4f2ea39ab6935c993) where runtime matters.

The output is a little bit confusing and needs some adjustment.

Could you make the computation into a command line toggle and produce something like this?
```
22235 watches in total
41 total instances for $USER
```

Currently it looks like this:
```
   22235  WATCHES TOTAL COUNT

      41  TOTAL INSTANCES COUNT

     INSTANCES  USER
     41         $USER
```

bdowling committed:
Updated script to collect inotify info once and use for all purposes

bdowling commented:
Because the bulk of my script was adding instance counts per process as well as per-user, it is a little hard to make that optional.

Instead I updated the PR though to collect the info once into a variable and re-process it to get the other unique stats needed, I also added a comment before the per-user instances table that I hope helps, as well as removing the INSTANCES TOTAL since that number has no real meaning, as INOTIFY is per-user limits.

This performed significantly faster on the larger systems I tested than the second find/grep/massage calls.

fatso83 commented:
The output is way better now, but the actual speed is still many times the older one, using about 6x as much kernel time and 2.5x the user time:
```
...
   42880  WATCHES TOTAL COUNT

INotify instances per user (e.g. limits specified by fs.inotify.max_user_instances): 

INSTANCES    USER
-----------  ------------------
50           carlerik
22           root
1            whoopsie
1            systemd-ti+
1            systemd-re+
1            messagebus
1            colord
1            avahi


real	0m1,802s
user	0m1,576s
sys	0m0,816s
```
(This was a `sudo` run, which is more taxing, as it includes other users' processes. The `x` numbers are relative to that)

But as you say: you fetch most of the important data in one call, and then most of the other time is probably spent just one sub-shell calls for doing all the data manipulation, so it is not necessarily that slow on big iron. 

I think it seems fine for now, and if Simon (which contributed the original speedup) or others find any perf issues we can revert it or work on improving it later on. Maybe we can find a way of passing that loop to `xargs` 😃

fatso83 commented:
Thanks!
